File: app/src/main/java/com/example/dagger2demojava/di/NetModule.java
package com.example.dagger2demojava.di;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;

@Module
public class NetModule {

    //告诉dagger可以通过该方法获得实例
    @Provides
    public Retrofit providerRetrofit(OkHttpClient okHttpClient) {
        return new Retrofit.Builder()
                .baseUrl("http://baidu.com/")
                .client(okHttpClient)
                .build();
    }

    @Provides
    public OkHttpClient providerOkHttpClient() {
        return new OkHttpClient();
    }

    //这种方式会自动找当前类中是否有提供该参数的方法，如果有的话就会调用默认的这个方法
    @Singleton
    @Provides
    public ApiService providerApiService(Retrofit retrofit) {
        return retrofit.create(ApiService.class);
    }
}

File: app06/src/main/java/com/example/app06/net/NetModule.kt
package com.example.app06.net

import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn

import dagger.hilt.components.SingletonComponent
import hu.akarnokd.rxjava3.retrofit.RxJava3CallAdapterFactory
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

@InstallIn(SingletonComponent::class)
@Module
class NetModule {


    @Provides
    fun providerUrl(): String {
        return "https://www.baidu.com/"
    }

    @Provides
    fun providerOkHttpClient(): OkHttpClient {
        return OkHttpClient()
            .newBuilder()
            .retryOnConnectionFailure(true)
            .build()
    }


    @Provides
    fun providerGsonConverterFactory(): GsonConverterFactory {
        return GsonConverterFactory.create()
    }

    @Provides
    fun providerRxJava3CallAdapterFactory(): RxJava3CallAdapterFactory {
        return RxJava3CallAdapterFactory.create()
    }


    @Provides
    fun providerRetrofit(
        url: String,
        okHttpClient: OkHttpClient,
        gsonConverterFactory: GsonConverterFactory,
        rxJava3CallAdapterFactory: RxJava3CallAdapterFactory
    ): Retrofit {

        return Retrofit.Builder()
            .baseUrl(url)
            .client(okHttpClient)
            .addConverterFactory(gsonConverterFactory)
            .addCallAdapterFactory(rxJava3CallAdapterFactory)
            .build()
    }

    @Provides
    fun providerApiService(retrofit: Retrofit): ApiService {
        return retrofit.create(ApiService::class.java)
    }
}